Opentracing Documentation (#5703)
matrix-org/synapse: docs/opentracing.rst

===========
OpenTracing
===========

Background
----------

OpenTracing is a semi-standard being adopted by a number of distributed tracing
platforms. It is a common api for facilitating vendor-agnostic tracing
instrumentation. That is, we can use the OpenTracing api and select one of a
number of tracer implementations to do the heavy lifting in the background.
Our current selected implementation is Jaeger.

OpenTracing is a tool which gives an insight into the causal relationship of
work done in and between servers. The servers each track events and report them

> to a centralised server - in our Synapse's case: Jaeger. The basic unit used to

Reviewer: ```suggestion
to a centralised server - in Synapse's case: Jaeger. The basic unit used to
```

finish.

Since this is undertaken in a distributed environment a request to another
server, such as an RPC or a simple GET, can be considered a span (a unit or
work) for the local server. This causal link is what OpenTracing aims to
capture and visualise. In order to do this metadata about the local server's
span, i.e the 'span context', needs to be included with the request to the
remote.

It is up to the remote server to decide what it does with the spans
it creates. This is called the sampling policy and it can be configured
through Jaeger's settings.

For OpenTracing concepts see 
https://opentracing.io/docs/overview/what-is-tracing/.

For more information about Jaeger's implementation see
https://www.jaegertracing.io/docs/

=====================
Seting up OpenTracing
=====================

To receive OpenTracing spans, start up a Jaeger server. This can be done
using docker like so:

.. code-block:: bash

   docker run -d --name jaeger
     -p 6831:6831/udp \
     -p 6832:6832/udp \
     -p 5778:5778 \
     -p 16686:16686 \
     -p 14268:14268 \
     jaegertracing/all-in-one:1.13

Latest documentation is probably at
https://www.jaegertracing.io/docs/1.13/getting-started/

> Enable opentracing in Synapse

Reviewer: ```suggestion
Enable OpenTracing in Synapse
```

-----------------------------

> Opentracing is not enabled by default. It must be enabled in the homeserver

Reviewer: ```suggestion
OpenTracing is not enabled by default. It must be enabled in the homeserver
```

config by uncommenting the config options under ``opentracing`` as shown in

> the [sample config](./sample_config.yaml). For example:

Reviewer: ```suggestion
the `sample config <./sample_config.yaml>`_. For example:
```

(I think)

.. code-block:: yaml

  opentracing:
    tracer_enabled: true
    homeserver_whitelist:
      - "mytrustedhomeserver.org"
      - "*.myotherhomeservers.com"

Homeserver whitelisting
-----------------------

> The homeserver whitelist is configured using regular expression. A list of regular

Reviewer: ```suggestion
The homeserver whitelist is configured using regular expressions. A list of regular
```

expressions can be given and their union will be compared when propagating any
spans contexts to another homeserver. 

Though it's mostly safe to send and receive span contexts to and from
untrusted users since span contexts are usually opaque ids it can lead to
two problems, namely:

- If the span context is marked as sampled by the sending homeserver the receiver will

>   sample it. Therefore two homeservers with wildly disparaging sampling policies

Reviewer: ```suggestion
  sample it. Therefore two homeservers with wildly different sampling policies
```

or diverging?

  could incur higher sampling counts than intended.

> - Span baggage can be arbitrary data. For safety this has been disabled in Synapse

Reviewer: This is the first time baggage is mentioned.

```suggestion
- Sending servers can attach arbitrary data to spans, known as 'baggage'. For safety this has been disabled in Synapse
```

===========

> Opentracing

Reviewer: This seems to be spelt `OpenTracing` in the upstream docs. Please can you use that consistently, rather than "Opentracing" / "opentracing"?

Reviewer: (likewise `synapse` -> `Synapse`, while I'm on naming)

===========

Background
----------

Opentracing is semi-standard being addopted by a number of distributed tracing

> platforms. It is a standardised api for facilitating vendor agnostic tracing

Author: A common api is good. OpenTracing has a nice intro stating what OpenTracing *isn't* but they never precisely pin down what OpenTracing *is*.

instrumentation. That is, we can use the opentracing api and select one of a
number of tracer implementations to do the heavy lifting in the background.
Our current selected implementation is Jaeger.

> Opentracing concepts can be found at

Reviewer: I'd find a very high-level overview of how it works useful here, to save me having to go and find the relevant bits in the upstream documentation. Just a sentence or two, like 

> Opentracing works by adding a header to HTTP requests, which identifies a 'span context', and whether it should be 'sampled'. This means that a single "span" can be traced between servers. It is up to the server what it does with the information it collects when it samples a span; in Synapse's case, it will report it to a Jaeger server.

(please edit for correctness!)

Author: I've changed this a lot.

https://opentracing.io/docs/overview/what-is-tracing/

> Python specific tracing concepts are at https://opentracing.io/guides/python/.

Reviewer: I'd like us to keep developer documentation separate from user (documentation). At the moment, most of this is stuff that is useful to know to someone who is running synapse, but not writing any code, but I think this paragraph is probably a bit unhelpful to a user rather than a developer.

Developer docs could end up in the same file, but it would need to be in a separate section, rather than mixed together. (Another option would be to put it in a docstring in `opentracing.py` or something.)

> Python specific tracing concepts are at https://opentracing.io/guides/python/.

Author: Is there an upper limit to the amount of documentation we put in a docstring? I haven't seen many long doc strings in the code base. Including the entire `Using opentracing in synapse` section in a docstring seems like a lot.

> Python specific tracing concepts are at https://opentracing.io/guides/python/.

Reviewer: There's no limit at all; people write whole books in there, especially for file-level docstrings.

> Python specific tracing concepts are at https://opentracing.io/guides/python/.

Author: The unsung sagas. I've just moved it to the docstring verbatim.

Note that synapse wraps opentracing in a small library in order to make the
opentracing dependency optional. That means that the access patterns are
different to those demonstrated here. However, it is still usefull to know.
Especially if opentracing is included as a full dependency in the future or if
you are modifying synapse's opentracing lib.

For more information about Jaeger's implementation see
https://www.jaegertracing.io/docs/

=================
Setup opentracing
=================

To receive opentracing spans start up a Jaeger server using docker like so

.. code-block:: bash

   docker run -d --name jaeger \ -e COLLECTOR_ZIPKIN_HTTP_PORT=9411 \

>      -p 5775:5775/udp \

Author: Only two of them are optional I'm afraid. I've removed those.

     -p 6831:6831/udp \
     -p 6832:6832/udp \
     -p 5778:5778 \
     -p 16686:16686 \
     -p 14268:14268 \
     -p 9411:9411 \
     jaegertracing/all-in-one:1.13

Latest documentation is probably at
https://www.jaegertracing.io/docs/1.13/getting-started/

Enable opentracing in synapse
-----------------------------

Opentracing is not enabled by default. It must be enabled in the homeserver
config by uncommenting the config options under ``opentracing``. For example:

.. code-block:: yaml

  opentracing:
    # Enable / disable tracer
    tracer_enabled: true
    # The list of homeservers we wish to expose our current traces to.
    # The list is a list of regexes which are matched against the
    # servername of the homeserver
    homeserver_whitelist:

>       - ".*"

Author: I'll give specific examples instead.

Homeserver whitelisting
-----------------------

The homeserver whitelist is configured using regex. A list of regexes can be
given and their union will be compared when propagating any spans through a
carrier. Most of the whitelist checks are encapsulated in the lib's injection

> and extraction method but be aware that using custom carriers or crossing

Reviewer: This feels like it's crossing over into developer documentation again. Suggest moving it elsewhere.

unchartered waters will require the enforcement of this whitelist.

``logging/opentracing.py`` has a ``whitelisted_homeserver`` method which takes
in a destination and compares it to the whitelist.

============================
Using opentracing in synapse
============================

Access to the opentracing api is mediated through the

> ``logging/opentracing.py`` lib. Opentracing is encapsulated such that

Reviewer: Is it really a library? more a module?

> no statefull spans from opentracing are used in synapses code. This allows

Reviewer: ```suggestion
no stateful spans from OpenTracing are used in Synapse's code. This allows
```

I'm not sure what stateful spans are, or why they would be used in Synapse's code. Perhaps that's just me being unfamiliar with some of the concepts though, in which case this is fine

Author: stateful was not a good adjective to use. The span objects created cannot be accessed by any other means than the module's methods.

opentracing to be easily disabled in synapse and thereby have opentracing as

> an optional dependency. This does however limit the number of modifyable spans

Reviewer: ```suggestion
an optional dependency. This does however limit the number of modifiable spans
```

> at any point in the code to one. From here out references to opentracing refer

Reviewer: > references to opentracing refer to the lib implemented in synapse

I can only find a couple of such references ([here](https://github.com/matrix-org/synapse/pull/5703/files#diff-7f16871d89a5bafeae6b6ec713bd857fR156) and [here](https://github.com/matrix-org/synapse/pull/5703/files#diff-7f16871d89a5bafeae6b6ec713bd857fR158)) outside code samples, and as far as I can tell they refer to the underlying opentracing library, so I'm not sure this what you mean.

Author: I meant references in the code snippets. I've clarified it.

to the lib implemented in synapse.

Tracing
-------

> In synapse it is not possible to start a non-active span. Spans can be started

Reviewer: inactive?

Author: This one is a little annoying. OpenTracing's definition of 'active' doesn't mean that the span is hasn't started or isn't relevant. It just means it's not directly accessible through opentracing's contextmanager. So, you can start a span which isn't inactive in the sense that it's disabled, hasn't started tracing or can't be logged to but is inactive because it hasn't been made the main span for this context. I don't know how to phrase this so that it's not confusing without a thorough understanding of OpenTracing.

> using the ``opentracing.start_active_span`` method. This returns a scope (see

Reviewer: this is probably clearer *without* the `opentracing.`, to distinguish from the real opentracing library.

opentracing docs) which is a context manager that needs to be entered and
exited. This is usually done by using ``with``.

.. code-block:: python

>    with start_active_span("operation name"):

Reviewer: you probably want a 

```
from synapse.logging.opentracing import start_active_span
```

here, for clarity

Reviewer: (similarly in the samples below)

       # Do something we want to tracer

Forgetting to enter or exit a scope will result in some mysterious grevious log
context errors.

At anytime where there is an active span ``opentracing.set_tag`` can be used to
set a tag on the current active span.

Tracing functions
-----------------

Functions can be easily traced using decorators. There is a decorator for
'normal' function and for functions which are actually deferreds. The name of
function becomes the operation name for the span.

.. code-block:: python

   # Start a span using 'normal_function' as the operation name
   @trace
   def normal_function(*args, **kwargs):
       # Does all kinds of cool and expected things
       return something_usual_and_useful

   # Start a span using 'deferred_function' as the operation name
   @trace_deferred

>    # Yes, there is a typo in the lib. I will fix this

Reviewer: there is?

Author: There was. Now I understand how quickly comments go out of date.

>    def deferred_function(*args, **kwargs):

Reviewer: should this have an `@defer.inlineCallbacks` too?

Author: yup

       # We start
       yield we_wait
       # we finish
       defer.returnValue(something_usual_and_useful)

Operation names can be explicitely set for functions by using
``trace_using_operation_name`` and
``trace_deferred_using_operation_name``

.. code-block:: python

   @trace_using_operation_name("A *much* better operation name")
   def normal_function(*args, **kwargs):
       # Does all kinds of cool and expected things
       return something_usual_and_useful

   @trace_deferred_using_operation_name("An operation name that fixes the typo!")
   # Yes, there is a typo in the lib. I will fix this
   def deferred_function(*args, **kwargs):
       # We start
       yield we_wait
       # we finish
       defer.returnValue(something_usual_and_useful)

Contexts and carriers
---------------------

There are a selection of wrappers for injecting and extracting contexts from

> carriers provided. Unfortunately opentracing's standard three are not adequate

Reviewer: standard three?

Author: The standard three injection methods. I've clarified.

> in the majority of cases. Also note that the binnary encoding format mandated

Reviewer: ```suggestion
in the majority of cases. Also note that the binary encoding format mandated
```

> by opentracing is not actually implemented by Jaeger and it will silently noop.

Author: Add a jaeger version
